### ISSUE TITLE
Fix cursor position model.

### ISSUE THREAD

**Issue by kisielk** [kisielk/vigo#38]
Cursors behave a bit inconsistently right now. Sometimes they return `\n`, other times they do not. Sometimes they allow the cursor to go past the end of line, other times they do not. This issue is to discuss the principles for cursor behaviour under various conditions.

**Comment by gchp** [comment]
Just did a quick test in vim. 

Normal mode - Pressing `$` brings me to the last character on the current line, not beyond.
Visual mode - Pressing `$` brings me one beyond the last character on the current line.

So it seems even vim has inconsistent behaviour ;)

**Comment by gchp** [comment]
My feeling is that `\n` should not be counted as part of the line. So given the following line of code:

```
func test() int {
```

`{` should be considered the last character in the line. At the moment, `len(cursor.Line.Data)` will return the length of the actual characters of the line, plus the newline char at the end. Thus it is possible to move the cursor over the newline and press `x` effectively joining the lines. I don't think this should be possible.

**Comment by kisielk** [comment]
It depends. If you are doing something like a regular expression or any other text processing such as counting characters, I'm pretty sure you need to count `\n`, and also consider it as a whitespace character.

In visual mode in vim, it's useful to be able to go beyond the end of the line because yanking either copies the newline or it doesn't.

**Comment by gchp** [comment]
So then cursors need some way of knowing whether or not to include that character. Simplified example:

``` go
func (c *Cursor) EOL(includeNewline bool) bool {
    l := len(c.Line.Data)
    if !includeNewline {
        l -= 1
    }
    return c.Boffset == l
}
```

Or make cursors have a concept of mode?

**Comment by kisielk** [comment]
I don't think it's necessary to put a boolean on every cursor, nor is it necessary for them to have modes. I think functions like EOL are actually fine if we take EOL to mean "the cursor is at the end of the line, on the newline character".

**Comment by gchp** [comment]
Fair enough. In that case it will be slight different to how vim handles it, albeit simpler. Which is fine by me :+1:

**Comment by kisielk** [comment]
How is it different than how Vim handles it?

**Comment by gchp** [comment]
In normal mode, vims cursor will not go the the newline character. It stops
at whatever is before it.

**Comment by kisielk** [comment]
Yes, we can still handle that in the mode. Move the cursor to the next character. If the next character is EOL, don't advance the visual cursor.

**Comment by gchp** [comment]
Ah, sorry, I misunderstood you earlier. I was thinking more about the
visual representation of this, rather than the functional bit underneath.
You're right.

Should vigo stick with the "religious" aspect of godit here then? ie that
newlines are always `\n` and not `\r\n`?

**Comment by kisielk** [comment]
I'd like to stick with `\n`. vigo is intended to be a programming editor and I think most developers use `\n`, even on windows these days (I think?)

**Comment by gchp** [comment]
I think so. Mostly anyways. I use Linux, but I work with some developers who use windows. Occasionally I will see some strange line endings sneaking around, but not often.

**Comment by kisielk** [comment]
It doesn't really matter actually, internally we can convert to `\n` when we read and just write it out that way. If later someone requests and option to write out with `\r\n` line endings instead, we can add that. However the internal representation will be `\n`, and a line ending will only count as one character.

**Comment by gchp** [comment]
Ok, cool.

So, what steps need to be taken on this issue then? I'll have a few hours free this evening to put together a patch, unless you've already got something on this?

**Comment by kisielk** [comment]
I haven't had time to think it all the way through just yet, so I'm not sure what exactly should be done.

**Comment by yarbelk** [comment]
Should we decouple the terminating character from our internal representation? from [wikipedia](http://en.wikipedia.org/wiki/Newline)

```
The Unicode standard defines a number of characters that conforming applications should recognize as line terminators:[3]

 LF:    Line Feed, U+000A
 VT:    Vertical Tab, U+000B
 FF:    Form Feed, U+000C
 CR:    Carriage Return, U+000D
 CR+LF: CR (U+000D) followed by LF (U+000A)
 NEL:   Next Line, U+0085
 LS:    Line Separator, U+2028
 PS:    Paragraph Separator, U+2029
```

This becomes really messy to look after

Personally - I think we should go for \n only for the moment, and I'm with @kisielk on mapping the buffer all to \n for our display purposes.
